Top / bottom setting for individual subtitles

artjomsR opened the pull request:
Mostly working for the extension and still needs cleaning up, but I wanted to ask first: is there any penalty for always having 2 subtitle containers (one at top and one at bottom), even if one of them is not used? E.g., all subtitles are displayed at the bottom?

killergerbah commented:
> is there any penalty for always having 2 subtitle containers (one at top and one at bottom), even if one of them is not used?

Sorry for the really late response on this. I probably clicked through this PR and then forgot to reply.
I don't think there should be a huge penalty. My guess is that the DOM cache will be the biggest drain on resources, but also probably not a big deal. But it would be good to avoid rendering the subtitles unnecessarily. E.g. if only top or only bottom then only render top or only render bottom.

artjomsR commented:
no problem, thanks for getting back to me! I've updated the PR now, let me know what you think :)

killergerbah commented on extension/src/controllers/subtitle-controller.ts:
```diff
@@ -113,36 +118,46 @@ export default class SubtitleController {
 
     cacheHtml() {
-        if (!(this.subtitlesElementOverlay instanceof CachingElementOverlay)) {
+        if (!(this.bottomSubtitlesElementOverlay instanceof CachingElementOverlay) || !(this.topSubtitlesElementOverlay instanceof CachingElementOverlay)) {
```
Currently, the check assumes that both overlays are always the same type, which may be true and enforced internally by this class.

But since it does not hurt, I think to be safe the `instanceof` check should be done independently for `bottomSubtitlesElementOverlay` and `topSubtitlesElementOverlay` before calling `uncacheHtml`.

killergerbah commented on extension/src/controllers/subtitle-controller.ts:
```diff
@@ -154,6 +169,22 @@ export default class SubtitleController {
         }
 
+        const newAlignments = allTextSubtitleSettings(newSubtitleSettings).map((s) => s.subtitleAlignment);
+        if (newAlignments !== this.subtitleTrackAlignments) {
```
This check will always return `true` because JavaScript will use object identity for comparison. This class already implements `_arrayEquals`, you can use that instead.

killergerbah commented on extension/src/controllers/subtitle-controller.ts:
```diff
@@ -169,21 +200,8 @@ export default class SubtitleController {
-
-    get subtitleAlignment() {
-        return this._subtitleAlignment;
+    getSubtitleTrackAlignment(trackIndex: number) {
```
Can be `private`

killergerbah commented on extension/src/controllers/subtitle-controller.ts:
```diff
@@ -451,7 +473,10 @@ export default class SubtitleController {
         }
 
-        this.subtitlesElementOverlay.dispose();
+        if (this.shouldRenderBottomOverlay)
```
Skip the `should*` checks - just dispose both overlays.

killergerbah commented on common/settings/settings.ts:
```diff
@@ -149,7 +150,8 @@ const subtitleSettingsKeysObject: { [key in keyof SubtitleSettings]: boolean } =
     subtitleBlur: true,
     imageBasedSubtitleScaleFactor: true,
-    subtitlePositionOffset: true,
+    bottomSubtitlePositionOffset: true,
```
Use `subtitlePositionOffset` for `bottomSubtitlePositionOffset` and add a comment saying it's the bottom offset. The naming is less clear but without a system to migrate the settings structure it's safer to preserve older names, since the extension and app share settings.

killergerbah reviewed (changes requested):
As always thank you. It took me a while to get to this because of work and also trying to get 1.5.0 out first.

artjomsR commented:
No problem for the delay, there's no rush :) 

I've updated the PR, should've addressed all your comments. You didn't address my previous comment, not sure if you have an opinion about this?

> I'm also thinking if we should remove the if checks from shouldRenderBottomOverlay (and remove it altogether), leaving only shouldRenderTopOverlay: boolean;: I think users would use either only bottom or bottom + top subtitles. But I don't see why someone would use only top subtitles

killergerbah commented:
> No problem for the delay, there's no rush :)
> 
> I've updated the PR, should've addressed all your comments. You didn't address my previous comment, not sure if you have an opinion about this?
> 
> > I'm also thinking if we should remove the if checks from shouldRenderBottomOverlay (and remove it altogether), leaving only shouldRenderTopOverlay: boolean;: I think users would use either only bottom or bottom + top subtitles. But I don't see why someone would use only top subtitles

Sorry forgot to reply to that as well. There are two reasons why I think we should leave it as is:

- It's not obvious to me how we would work that into the current settings UI. For example you would have to prevent users from switching all tracks to the top. It would seem then the track specific settings would have to be aware of each other which might introduce complexity.
- There are users who use top only. For example to add second language subtitles to websites that already have their own subtitles.

artjomsR commented:
I meant to keep the functionality as is, simply skip the `if shouldRenderBottomOverlay` check and always invoke all functionality related to the bottom overlay. Once again, based on the principle that _most_ people would use the bottom container. 

But as you say, if considerable number of users use top only, then we can leave everything as is :)

killergerbah commented on scripts/anki-connect-proxy/anki-connect-proxy.exe:
I think this file should be part of a different commit.

killergerbah reviewed (commented):
It looks like you removed the `subtitleAlignment` setting. If you're not using that setting, how are you intending for users to be able to pick the alignment for individual tracks?

artjomsR commented:
> It looks like you removed the `subtitleAlignment` setting. If you're not using that setting, how are you intending for users to be able to pick the alignment for individual tracks?

in common/settings/settings.ts, I've moved it from SubtitleSettings (which I understood as "global" settings) to TextSubtitleSettings (which I understood as settings which apply to individual tracks), which is effectively all this PR is about 

Uploading Replay_2024-10-12_10-26-57.mp4…

killergerbah commented:
Understood!

killergerbah commented on extension/src/controllers/subtitle-controller.ts:
```diff
@@ -175,21 +208,8 @@ export default class SubtitleController {
-    get subtitleAlignment() {
-        return this._subtitleAlignment;
+    private _getSubtitleTrackAlignment(trackIndex: number) {
+        return this.subtitleTrackAlignments[trackIndex] || 'bottom';
```
As mentioned above the default value should really be the setting for the 0th track:

```
this.subtitleTrackAlignments[trackIndex] || this.subtitleTrackAlignments[0]
```

artjomsR commented:
I've added your suggested change of `|| this.subtitleTrackAlignments[0]` which seems to have fixed the bug for me, I've tested several times. So I'm not sure if the caching changes are still required now

killergerbah commented:
Sorry, yeah you're right. Thanks again for your patience.
